group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java

```diff
@@ -3022,143 +3242,2087 @@ public void testOnLoaded() {
+        assertTrue(responseFuture.isDone());
+        assertEquals(Errors.COORDINATOR_NOT_AVAILABLE.code(), responseFuture.get(5, TimeUnit.SECONDS).errorCode());
+    }
+
+    @Test
+    public void testJoinGroupProtocolTypeIsNotProvidedWhenAnErrorOccurs() throws Exception {
+        GroupMetadataManagerTestContext context = new GroupMetadataManagerTestContext.Builder()
+            .build();
+
+        JoinGroupRequestData request = new JoinGroupRequestBuilder()
+            .withGroupId("group-id")
+            .withMemberId("member-id")
+            .withDefaultProtocolTypeAndProtocols()
+            .build();
+
+        CompletableFuture<JoinGroupResponseData> responseFuture = new CompletableFuture<>();
+        CoordinatorResult<Void, Record> result = context.sendGenericGroupJoin(request, responseFuture);
+
+        assertTrue(result.records().isEmpty());
+        assertTrue(responseFuture.isDone());
+        assertEquals(Errors.UNKNOWN_MEMBER_ID.code(), responseFuture.get().errorCode());
+        assertNull(responseFuture.get().protocolType());
+    }
+
+    @Test
+    public void testJoinGroupReturnsTheProtocolType() throws Exception {
+        GroupMetadataManagerTestContext context = new GroupMetadataManagerTestContext.Builder()
+            .build();
+        context.createGenericGroup("group-id");
+
+        // Leader joins
+        JoinGroupRequestData request = new JoinGroupRequestBuilder()
+            .withGroupId("group-id")
+            .withMemberId(UNKNOWN_MEMBER_ID)
+            .withDefaultProtocolTypeAndProtocols()
+            .build();
+
+        CompletableFuture<JoinGroupResponseData> leaderResponseFuture = new CompletableFuture<>();
+        CoordinatorResult<Void, Record> result = context.sendGenericGroupJoin(request, leaderResponseFuture);
+
+        assertTrue(result.records().isEmpty());
+        assertFalse(leaderResponseFuture.isDone());
+
+        // Member joins
+        CompletableFuture<JoinGroupResponseData> memberResponseFuture = new CompletableFuture<>();
+        result = context.sendGenericGroupJoin(request, memberResponseFuture);
+
+        assertTrue(result.records().isEmpty());
+        assertFalse(memberResponseFuture.isDone());
+
+        // Complete join group phase
+        context.sleepAndAssertEmptyResult(context.genericGroupInitialRebalanceDelayMs);
+        assertTrue(leaderResponseFuture.isDone());
+        assertTrue(memberResponseFuture.isDone());
+
+        assertEquals(Errors.NONE.code(), leaderResponseFuture.get(5, TimeUnit.SECONDS).errorCode());
+        assertEquals("consumer", leaderResponseFuture.get(5, TimeUnit.SECONDS).protocolType());
+        assertEquals(Errors.NONE.code(), memberResponseFuture.get(5, TimeUnit.SECONDS).errorCode());
+        assertEquals("consumer", memberResponseFuture.get(5, TimeUnit.SECONDS).protocolType());
+    }
+
+    @Test
+    public void shouldDelayInitialRebalanceByGroupInitialRebalanceDelayOnEmptyGroup() throws Exception {
```

Comment:
nit: While we are here, would it make to normalize the name of all tests starting with `should`? They should ideally start with `test...` like the others.